src/Workspaces/Core/Portable/Workspace/Solution/CachedVersionTextSource.cs

```diff
@@ -0,0 +1,47 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Internal.Log;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis
+{
+    /// <summary>
+    /// A text source that keeps the text version cached.
+    /// </summary>
+    internal class CachedVersionTextSource : ValueSource<TextAndVersion>, ITextVersionable
```

Comment:
That still implies API change, as we'd also need to supply a CancellationToken, and all current callers would need to change to call it.

As it stands, TextAndVersion is just a simple immutable data structure, just like DocumentInfo, etc. This problem doesn't need to be solved by changing it.  The problem is totally contained within the DocumentState, so it is better to just implement the solution there.